### DISCUSSION TITLE
Trouble with Displaying the Correct 'Hour' value

### DISCUSSION THREAD

**Discussion by farTooOld** [micropython/micropython#9196] [Using MicroPython]
OSX 12.5
MicroPython v1.19.1
Thonny 3.3 13
Python 3.7.9
Pi Pico with RP2040

Problem: Printing / Displaying Numbers from Time Codes

I come to Python with over 40 years of programming experience and never had issues with Using/Calculating, Printing/Displaying numbers (in all other programming languages). That is to say, once understanding the number type and formatting particulars for it with respect to Display type (i.e., Monitor, LCD, TFT...)

I'm having trouble with Displaying the correct 'Hour' value, as follows:

Goal: Using 24hr clock, I Grab the 'Hour' from ds3231 and add '7' to it to yield GMT. The current ds3231 Hour is set to my local time and my time zone is 7hr behind GMT). Rather than setting for GMT and subtracting 7hr to get local, I prefer to set for local time and add 7hr to get GMT)

• I can get the local HR and display it, no problem
• The local time (including HR) is always correct
• For GMT's Hr, logic suggests simply adding 7hr to the local hr BUT...

... the 'But' is this:
• All is good from 12am (0am) through 9am local. Then, at 10am local time, the GMThr jumps from 17 to 23 (a delta of 13hrs! Not, 7hrs) and, when local time is 20hrs, GMThr jumps from 19hrs ahead of local to 39).

Perhaps best to look at the printout, below

The solution's I tried:
• Different number formats, Int, Str, i, d, x, s, r, f
• I've tried using Bytes, Ints, Floats, etc...
• I've printed out the 'number' types example: type (print(type(GMTHr)) and they show the correct type

No matter what combinations I use, I never get a consistent addition of 7 to the local Hr.  What I'm I doing wrong, or even better, what should I do to both understand this problem and fix it?

**Printout of a partial day's time data**

```
01:00:00       08:00:00 
02:00:00       09:00:00 
03:00:00       10:00:00
04:00:00       11:00:00 
05:00:00       12:00:00 
06:00:00       13:00:00 
07:00:00       14:00:00 
08:00:00       15:00:00 
09:00:00       16:00:00 
10:00:00       23:00:00
11:00:00       24:00:00 
12:00:00       25:00:00 
13:00:00       26:00:00 
14:00:00       27:00:00 
15:00:00       28:00:00 
16:00:00       29:00:00 
17:00:00       30:00:00 
18:00:00       31:00:00 
19:00:00       32:00:00 
20:00:00       39:00:00 
21:00:00       40:00:00 
22:00:00       41:00:00  
23:00:00       42:00:00 
00:00:00       07:00:00
```

**Portion of the Relevant Code:**
```py
while True:
 #------------------------- TIME -----------------------------------------------
    t = ds.read_time()             # read the real RTC  

    iwonder = 7               
    print(iwonder)
    
    tofu = int(t[2] + iwonder)
    gmtHR = tofu
    
 #------------------------- TIME -----------------------------------------------
    GMTtime = "%02d:%02x:%02x" %(gmtHR,t[1],t[0])
    print('gmt:', GMTtime)  # yes!
    
   #===test
    GMTtime_i = "%02i:%02x:%02x" %(gmtHR,t[1],t[0])
    print('gmt: i', GMTtime_i)  
    
    GMTtime_x = "%02x:%02x:%02x" %(gmtHR,t[1],t[0])
    print('gmt: x', GMTtime_x) 
    
    GMTtime_s = "%02s:%02x:%02x" %(gmtHR,t[1],t[0]) 
    print('gmt: s', GMTtime_s) 
    
    GMTtime_r = "%02r:%02x:%02x" %(gmtHR,t[1],t[0]) 
    print('gmt: r', GMTtime_r)  
    
    GMTtime_f = "%02.0f:%02x:%02x" %(gmtHR,t[1],t[0]) 
    print('gmt: f', GMTtime_f) 
    
    print("\n")
   #========
    
    LocDSTtime = "%02x:%02x:%02x" %(t[2],t[1],t[0])
    print("DST:",LocDSTtime)     
 #------------------------- DATE -----------------------------------------------
    LocDate = "%02x/%02x/20%x" %(t[5],t[4],t[6])  # Re-arrange to Put 20 in front of yr
    GMTDate = "%02x/%02x/20%x" %(t[5],t[4],t[6])  # Re-arrange to Put 20 in front of yr
 
    print("Local Date: %02x/%02x/20%x" %(t[5],t[4],t[6]), " DST: %02x:%02x:%02x" %(t[2],t[1],t[0]))
    print("GMT Date:   %02x/%02x/20%x" %(t[5],t[4],t[6]), " GMT: %02d:%02x:%02x" %(gmtHR,t[1],t[0]))                                                              # use 'x'
 #------------------------- DAY LABEL ------------------------------------------
    # this Day is from the RTC
    LocdayLabel = 'Loc Day Error?'   
    if t[3] == 1: LocdayLabel = "Monday"
    if t[3] == 2: LocdayLabel = "Tuesday"
    if t[3] == 3: LocdayLabel = "Wednesday"
    if t[3] == 4: LocdayLabel = "Thursday"
    if t[3] == 5: LocdayLabel = "Friday"
    if t[3] == 6: LocdayLabel = "Saturday"
    if t[3] == 7: LocdayLabel = "Sunday"
    print(t[3], '   '+LocdayLabel)
   
    if (t[1] ==0 and t[0]==0):
        display.clear()
        print('Local Cleared',t[2], t[1])
    
    # this Day is from the New 'GMT time/date'
    GMTdayLabel = 'Loc Day Error?'
    if t[3] == 1: GMTdayLabel = "Monday"
    if t[3] == 2: GMTdayLabel = "Tuesday"
    if t[3] == 3: GMTdayLabel = "Wednesday"
    if t[3] == 4: GMTdayLabel = "Thursday"
    if t[3] == 5: GMTdayLabel = "Friday"
    if t[3] == 6: GMTdayLabel = "Saturday"
    if t[3] == 7: GMTdayLabel = "Sunday"
    print(t[3], '   '+GMTdayLabel)
   
    if (t[1] ==0 and t[0]==0):
        display.clear() # clear yesterday's screen
        print('GMT Cleared',t[2], t[1])
        
 #------------------------------ TFT DISPLAY -----------------------------------
    display.draw_text(0, 10, "GMT: "+GMTtime, unispace, color565(255, 255, 255))  
    display.draw_text(0, 40, "DST: "+LocDSTtime, unispace, color565(255, 255, 255))   
        
    display.draw_text(0, 70, "Loc Date: "+LocDate, unispace, color565(0, 255, 255))
    display.draw_text(0, 100, "GMT Date: "+GMTDate, unispace, color565(0, 255, 255))
    
    display.draw_text(0, 130, "Loc Day: "+LocdayLabel, unispace, color565(0, 255, 255))
    display.draw_text(0, 160, "GMT Day: "+GMTdayLabel, unispace, color565(0, 255, 255)) 
    
    reading = sensor_temp.read_u16() * conv
    temperature = 27 - (reading -0.706)/0.001721
    theTemp = str('%.1f' % (temperature * (9/5) + 32))
    display.draw_text(0, 190, "Temp: "+ theTemp, unispace, color565(255, 255, 0))`
```

(Edit by @jimmo -- add code formatting using triple-backticks `` ``` ``).

**Comment by cwalther** [comment]
I’m having some trouble following your code because it’s very long and oddly formatted and contains a lot more `print`s than the output you quote seems to show. Could you perhaps strip it down to the part that produces the quoted output? And could you also output what the `t` is that you are reading from the RTC at the top? It looks like it might be a tuple or a `struct_time` or something like that. I’m not familiar with DS3231 drivers and what exact data format they output. 

That said, the output looks like something is interpreting the hour as hexadecimal. E.g. in the row `22:00:00 41:00:00`, 0x22 = 34, and adding 7 to that makes 41 (decimal).

**Reply by farTooOld** [reply]
Thanks for looking...

The 't' is a tuple from the ds3132 (very common)

Yes, I know Hex and I've also tried adding Hex... no matter what/how I do it, what should work doesn't. That's why I posted the various Print lines...  Below is the salient code for printing the Hr in different formats

The  t[2]  is the Hr from the ds3231's tuple

```
 #------------------------- TIME -----------------------------------------------
    t = ds.read_time()             # read the real RTC  

    iwonder = 7               
    print(iwonder)
    
    tofu = int(t[2] + iwonder)
    gmtHR = tofu
    
 #------------------------- TIME print -----------------------------------------
    GMTtime = "%02d:%02x:%02x" %(gmtHR,t[1],t[0])
    print('gmt:', GMTtime)  # yes!
    
   #===test
    GMTtime_i = "%02i:%02x:%02x" %(gmtHR,t[1],t[0])
    print('gmt: i', GMTtime_i)  # yes!
    
    GMTtime_x = "%02x:%02x:%02x" %(gmtHR,t[1],t[0])
    print('gmt: x', GMTtime_x)  # yes!
    
    GMTtime_s = "%02s:%02x:%02x" %(gmtHR,t[1],t[0]) 
    print('gmt: s', GMTtime_s)  # yes!
    
    GMTtime_r = "%02r:%02x:%02x" %(gmtHR,t[1],t[0]) 
    print('gmt: r', GMTtime_r)  # yes!
    
    GMTtime_f = "%02.0f:%02x:%02x" %(gmtHR,t[1],t[0]) 
    print('gmt: f', GMTtime_f)  # yes!
    
    print("\n")
   #========
    
    LocDSTtime = "%02x:%02x:%02x" %(t[2],t[1],t[0])
    print("DST:",LocDSTtime)     

```

**Reply by cwalther** [reply]
This is obviously not the code that produced the output in question – e.g. most prints here start with “gmt:”, which is nowhere to be seen in the quoted output. So, can you post the exact output of this code (assuming it shows the effect that you are confused about)? Also, I’d like to see the exact output of `print(t)`, and not just your colloquial description of it, so that I know what the data type (and content) of `t[2]` is.

**Reply by farTooOld** [reply]
The Code I posted in first post is the Complete Code (except for the I2C, SPI and Import declaration). That's why I posted it, so there wouldn't be anything needed beyond further explanation.

So, given you don't need the I2C, SPI and Import's list, you got the full Code

All those 'Print' lines are for my debugging this.

Here's a print of 't' for one moment in time
t =  b'\x14\x17\x14\x05\x02\t"'
<class 'bytes'>

The  ds3231 info
To Set the Time, these two lines are used (and, no problem reading Time & Date. Only problem is Adding 7 to the Hr field. And, though this posted code doesn't show my various attempts at different ways of Adding 7,  I did try breaking into individual Int, Hex, Byte, Float...etc.  Nothing solved it. I am sure it's 'me' I just don't see the problem):

#current_time = b'\x00\x52\x12\x02\x18\x05\x21' # sec min hour week day mon year
#ds.set_time(current_time)

EDIT:
re [ “gmt:”, which is nowhere " ]  gmt (lowercase is only a Text  for what follows, which will be either gmtHR (or gmtDate). There is No actual variable 'gmt'

**Reply by cwalther** [reply]
> `t = b'\x14\x17\x14\x05\x02\t"'`

Ah, see, that’s not a tuple, that’s a `bytes`. That doesn’t explain the problem though, because `t[2]` is still an `int` as one would expect.

I still don’t see anything in your “Complete Code” that would produce output like
```
01:00:00 08:00:00
02:00:00 09:00:00
…
23:00:00 42:00:00
00:00:00 07:00:00
```
so I’m sorry, I can’t reproduce (and therefore can’t explain) your problem.

**Reply by farTooOld** [reply]
First of all, Thank You. 

You confirmed there is a problem outside of my code (though you didn't exactly say it and have seen only my posted code).

I suspect it's the ds3231 Library I'm using or perhaps Thonny (I've had problems with it). I'll dig into it...

Thanks again!

**Reply by jimmo** [reply]
@farTooOld I have updated your post to format the code snippet correctly -- see the (updated) instructions on formatting at [the welcome post](https://github.com/orgs/micropython/discussions/9111).

I'm a bit confused trying to understand what's going on, and unsure why hexadecimal needs to be involved in any part of this.

Which ds3231 library are you using? I would recommend Peter's one here: https://github.com/peterhinch/micropython-samples/tree/master/DS3231

Note: Just a random suspicion about the jump at 10am going to 23 when 7 is added... 23-7 is 16, which is how 10 would be represented in BCD. So I'm going to guess that binary-coded-decimal is probably involved somewhere. So 10 would be represented as the byte 0x1 0x0 (i.e. 16 decimal). Hence why you see 23.

So if you want to continue using whatever library you're using, you're going to need to "un-BCD" the bytes.

```py
def unbcd(x):
  return ((x >> 4) & 0xf) * 10 + (x & 0xf)

t = bytes([0, 16, 0, 0])  # This is what 10am would look like
hour = unbcd(t[1]) + 7
print(hour) # prints 17
```

**Reply by jimmo** [reply]
Ahh and this is why you're printing in everything in hexadecimal -- when you print a BCD-encoded number as hexadecimal, you get it as a decimal string.

**Reply by cwalther** [reply]
Ah, good point – BCD-encoded original data would indeed explain things. Thanks jimmo!

**Reply by farTooOld** [reply]
I totally forgot about BCD (out of sight =  out of mind)...

I'll fix the code to confirm handling the BCD takes care of it. Then, use your recommended ds3231 lib which may have additional features...

Thank you 'all' for the help, much appreciated!
